File: lib/spells/effects/UnitEffect.cpp
/*
 * UnitEffect.cpp, part of VCMI engine
 *
 * Authors: listed in file AUTHORS in main folder
 *
 * License: GNU General Public License v2.0 or later
 * Full text of license available in license.txt file, in main folder
 *
 */
#include "StdInc.h"

#include "UnitEffect.h"

#include "../ISpellMechanics.h"

#include "../../battle/CBattleInfoCallback.h"
#include "../../battle/Unit.h"
#include "../../serializer/JsonSerializeFormat.h"
#include <vcmi/spells/Spell.h>

VCMI_LIB_NAMESPACE_BEGIN

namespace spells
{
namespace effects
{

void UnitEffect::adjustAffectedHexes(BattleHexArray & hexes, const Mechanics * m, const Target & spellTarget) const
{
	for(const auto & destnation : spellTarget)
		hexes.insert(destnation.hexValue);
}

bool UnitEffect::applicableGeneral(Problem & problem, const Mechanics * m) const
{
	auto mainFilter = [this, m](const battle::Unit * unit){ return applicableUnit(m, unit, false, false);};

	//stack effect is applicable in general if there is at least one smart target
	auto targets = m->battle()->battleGetUnitsIf(mainFilter);

	if(targets.empty())
		return m->adaptProblem(ESpellCastProblem::NO_APPROPRIATE_TARGET, problem);

	return true;
}

bool UnitEffect::applicableTarget(Problem & problem, const Mechanics * m, const Target & target) const
{
	//stack effect is applicable if it affects at least one target (smartness should not be checked)
	//assume target correctly transformed, just reapply filter

	for(const auto & item : target)
		if(item.unitValue)
			if(applicableUnit(m, item.unitValue, false, true))
				return true;

	return false;
}

bool UnitEffect::applicableUnit(const Mechanics * m, const battle::Unit * unit, bool alwaysSmart, bool alwaysReceptive) const
{
	if (!isValidTarget(m, unit))
		return false;

	if (!alwaysReceptive && !isReceptive(m, unit))
		return false;

	const bool smart = m->isSmart() || alwaysSmart;
	const bool ignoreOwner = !smart;
	return ignoreOwner || m->ownerMatches(unit);
}

Target UnitEffect::filterTarget(const Mechanics * m, const Target & target) const
{
	Target res;
	vstd::copy_if(target, std::back_inserter(res), [this, m](const Destination & d)
	{
		return d.unitValue && isValidTarget(m, d.unitValue) && isReceptive(m, d.unitValue);
	});
	return res;
}

Target UnitEffect::transformTarget(const Mechanics * m, const Target & aimPoint, const Target & spellTarget) const
{
	if(chainLength > 1)
		return transformTargetByChain(m, aimPoint, spellTarget);
	else
		return transformTargetByRange(m, aimPoint, spellTarget);
}

Target UnitEffect::transformTargetByRange(const Mechanics * m, const Target & aimPoint, const Target & spellTarget) const
{
	auto mainFilter = [this, m](const battle::Unit * unit){ return applicableUnit(m, unit, false, false);};

	Target spellTargetCopy(spellTarget);

	// make sure that we have valid target with valid aim, even if spell have invalid range configured
	// TODO: check than spell range is actually not valid
	// also hackfix for banned creature massive spells
	// FIXME: potentially breaking change: aimPoint may NOT be in Target - example: frost ring
	if(!aimPoint.empty() && spellTarget.empty())
		spellTargetCopy.insert(spellTargetCopy.begin(), Destination(aimPoint.front()));

	std::set<const battle::Unit *> targets;

	if(m->isMassive())
	{
		//ignore spellTarget and add all stacks
		auto units = m->battle()->battleGetUnitsIf(mainFilter);
		for(const auto *unit : units)
			targets.insert(unit);
	}
	else
	{
		//process each tile
		for(const Destination & dest : spellTargetCopy)
		{
			if(dest.unitValue)
			{
				if(mainFilter(dest.unitValue))
					targets.insert(dest.unitValue);
			}
			else if(dest.hexValue.isValid())
			{
				//select one unit on tile, prefer alive one
				const battle::Unit * targetOnTile = nullptr;

				auto predicate = [&](const battle::Unit * unit)
				{
					return unit->coversPos(dest.hexValue) && mainFilter(unit);
				};

				auto units = m->battle()->battleGetUnitsIf(predicate);

				for(const auto *unit : units)
				{
					if(unit->alive())
					{
						targetOnTile = unit;
						break;
					}
				}

				if(targetOnTile == nullptr && !units.empty())
					targetOnTile = units.front();

				if(targetOnTile)
					targets.insert(targetOnTile);
			}
			else
			{
				logGlobal->debug("Invalid destination in spell Target");
			}
		}
	}

	if(m->alwaysHitFirstTarget())
	{
		//TODO: examine if adjustments needed related to INVINCIBLE bonus
		if(!aimPoint.empty() && aimPoint.front().unitValue)
			targets.insert(aimPoint.front().unitValue);
	}

	Target effectTarget;

	for(const auto *s : targets)
		effectTarget.push_back(Destination(s));

	return effectTarget;
}

Target UnitEffect::transformTargetByChain(const Mechanics * m, const Target & aimPoint, const Target & spellTarget) const
{
	Target byRange = transformTargetByRange(m, aimPoint, spellTarget);

	if(byRange.empty())
	{
		return Target();
	}

	const Destination & mainDestination = byRange.front();

	if(!mainDestination.hexValue.isValid())
	{
		return Target();
	}

	BattleHexArray possibleHexes;

	auto possibleTargets = m->battle()->battleGetUnitsIf([&](const battle::Unit * unit) -> bool
	{
		return isReceptive(m, unit) && isValidTarget(m, unit);
	});

	for(const auto *unit : possibleTargets)
	{
		for(const auto & hex : unit->getHexes())
			possibleHexes.insert(hex);
	}

	BattleHex destHex = mainDestination.hexValue;
	Target effectTarget;

	for(int32_t targetIndex = 0; targetIndex < chainLength; ++targetIndex)
	{
		const auto *unit = m->battle()->battleGetUnitByPos(destHex, true);

		if(!unit)
			break;

		bool wouldResist = m->wouldResist(unit);
		if(m->alwaysHitFirstTarget() && targetIndex == 0)
			effectTarget.emplace_back(unit);
		if(wouldResist && targetIndex == 0)
		{
			// if first target resists, chain ends here, resistance animation played
			effectTarget.emplace_back(unit);
			break;
		}
		else if(isReceptive(m, unit) && isValidTarget(m, unit) && !wouldResist)
			effectTarget.emplace_back(unit);
		else if(isReceptive(m, unit) && isValidTarget(m, unit) && wouldResist)
		{
			// target is skipped, no magic resistance animation (Heroes 3 logic)
			targetIndex--;
		}
		else
			effectTarget.emplace_back();

		for(const auto & hex : unit->getHexes())
			if (possibleHexes.contains(hex))
				possibleHexes.erase(hex);

		if(possibleHexes.empty())
			break;

		destHex = BattleHex::getClosestTile(unit->unitSide(), destHex, possibleHexes);
	}

	return effectTarget;
}

bool UnitEffect::isValidTarget(const Mechanics * m, const battle::Unit * unit) const
{
	// TODO: override in rising effect
	// TODO: check absolute immunity here

	return unit->isValidTarget(false);
}

bool UnitEffect::isReceptive(const Mechanics * m, const battle::Unit * unit) const
{
	if (unit->isInvincible() && m->isNegativeSpell())
		return false;
	if(ignoreImmunity)
	{
		//ignore all immunities, except specific absolute immunity(VCMI addition)
		return !unit->hasAbsoluteImmunity(m->getSpellId());
	}
	else
	{
		return m->isReceptive(unit);
	}
}

void UnitEffect::serializeJsonEffect(JsonSerializeFormat & handler)
{
	handler.serializeBool("ignoreImmunity", ignoreImmunity);
	handler.serializeInt("chainLength", chainLength, 0);
	handler.serializeFloat("chainFactor", chainFactor, 0);
	serializeJsonUnitEffect(handler);
}


}
}

VCMI_LIB_NAMESPACE_END

File: scripts/unitEffect.lua
local Script = {}
Script.__index = Script
Script.type = "unitEffect"

--- Actually casts the spell and applies all changes caused by it.
--- Use `server` parameter to apply changes on specified target(s).
--- target contains units pre-filtered by filterTarget,
--- so every dest.unit is guaranteed alive and receptive.
function Script:apply(mechanics, server, target)
	return
end

--- Returns true if at least one valid target exists on the battlefield.
function Script:applicableGeneral(mechanics, problem)
	local found = mechanics:getBattle():getUnitsIf(function(unit)
		if not self:isValidTarget(mechanics, unit) then return false end
		if not self:isReceptive(mechanics, unit) then return false end
		if mechanics:isSmart() and not mechanics:ownerMatches(unit) then return false end
		return true
	end)
	if #found == 0 then
		problem:addStandard(mechanics, ENUM.SpellCastProblem.noAppropriateTarget)
		return false
	end
	return true
end

--- Returns true if at least one unit in target is valid (ignoring receptiveness per H3 rules).
function Script:applicableTarget(mechanics, problem, target)
	for _, dest in ipairs(target) do
		local unit = dest.unit
		if unit and self:isValidTarget(mechanics, unit) then
			if not mechanics:isSmart() or mechanics:ownerMatches(unit) then
				return true
			end
		end
	end
	return false
end

--- Returns true if `unit` can be affected by this spell effect.
function Script:isValidTarget(mechanics, unit)
	return unit:isValidTarget(false)
end

--- Returns true if `unit` is receptive to this spell effect (not immune).
function Script:isReceptive(mechanics, unit)
	if unit:isInvincible() and mechanics:isNegative() then
		return false
	end
	if self.ignoreImmunity then
		return not unit:hasAbsoluteImmunity(mechanics:getSpell())
	else
		return mechanics:isReceptive(unit)
	end
end

--- Returns health change preview data for hover tooltip.
function Script:getHealthChange(mechanics, spellTarget)
	return { hpDelta = 0, unitsDelta = 0 }
end

--- Adjusts the spell's required target types. Return the (possibly modified) types array.
function Script:adjustTargetTypes(mechanics, types)
	return types
end

--- Adjusts affected hexes for visual highlighting. Inserts hex positions from spellTarget.
function Script:adjustAffectedHexes(mechanics, hexes, spellTarget)
	for _, dest in ipairs(spellTarget) do
		if dest.hex and dest.hex:isValid() then
			hexes:insert(dest.hex)
		end
	end
end

--- Filters the transformed target, keeping only valid and receptive units.
function Script:filterTarget(mechanics, target)
	local result = {}
	for _, dest in ipairs(target) do
		local unit = dest.unit
		if unit and self:isValidTarget(mechanics, unit) and self:isReceptive(mechanics, unit) then
			table.insert(result, dest)
		end
	end
	return result
end

--- Builds the full list of affected units from the spell target.
--- Dispatches to transformByChain or transformByRange based on chainLength parameter.
function Script:transformTarget(mechanics, aimPoint, spellTarget)
	local chainLength = self.chainLength or 0
	if chainLength > 1 then
		return self:transformByChain(mechanics, aimPoint, spellTarget, chainLength)
	else
		return self:transformByRange(mechanics, aimPoint, spellTarget)
	end
end

--- Resolves the effective spell target from hex/unit destinations (non-chain spells).
--- Mirrors C++ UnitEffect::transformTargetByRange.
function Script:transformByRange(mechanics, aimPoint, spellTarget)
	local battle = mechanics:getBattle()

	local function mainFilter(unit)
		if not self:isValidTarget(mechanics, unit) then return false end
		if not self:isReceptive(mechanics, unit) then return false end
		if mechanics:isSmart() and not mechanics:ownerMatches(unit) then return false end
		return true
	end

	-- ensure we have a non-empty target to work from
	local effectiveTarget = spellTarget
	if #aimPoint > 0 and #spellTarget == 0 then
		effectiveTarget = { aimPoint[1] }
	end

	local targetsSet = {}
	local targetsOrder = {}

	local function addUnit(unit)
		if not targetsSet[unit] then
			targetsSet[unit] = true
			table.insert(targetsOrder, unit)
		end
	end

	if mechanics:isMassive() then
		local units = battle:getUnitsIf(mainFilter)
		for _, unit in ipairs(units) do
			addUnit(unit)
		end
	else
		for _, dest in ipairs(effectiveTarget) do
			if dest.unit then
				if mainFilter(dest.unit) then
					addUnit(dest.unit)
				end
			elseif dest.hex and dest.hex:isValid() then
				local onTile = battle:getUnitsIf(function(unit)
					return unit:coversPos(dest.hex) and mainFilter(unit)
				end)
				local chosen = nil
				for _, unit in ipairs(onTile) do
					if unit:isAlive() then
						chosen = unit
						break
					end
				end
				if chosen == nil and #onTile > 0 then
					chosen = onTile[1]
				end
				if chosen then
					addUnit(chosen)
				end
			end
		end
	end

	if mechanics:alwaysHitFirstTarget() and #aimPoint > 0 and aimPoint[1].unit then
		addUnit(aimPoint[1].unit)
	end

	local result = {}
	for _, unit in ipairs(targetsOrder) do
		table.insert(result, { unit = unit, hex = unit:getPosition() })
	end
	return result
end

--- Resolves chain-lightning style targets (chain of nearest units).
--- Mirrors C++ UnitEffect::transformTargetByChain.
function Script:transformByChain(mechanics, aimPoint, spellTarget, chainLength)
	local byRange = self:transformByRange(mechanics, aimPoint, spellTarget)
	if #byRange == 0 then return {} end

	local mainDest = byRange[1]
	if not mainDest.hex or not mainDest.hex:isValid() then return {} end

	local battle = mechanics:getBattle()
	local alwaysHitFirst = mechanics:alwaysHitFirstTarget()
	local effectTarget = {}
	local processedIds = {}

	local function isEligible(unit)
		return self:isReceptive(mechanics, unit) and self:isValidTarget(mechanics, unit)
	end

	local function buildPossibleHexes()
		local units = battle:getUnitsIf(function(u)
			return isEligible(u) and not processedIds[u]
		end)
		if #units == 0 then return nil end
		local hexes = units[1]:getHexes()
		for i = 2, #units do
			local uHexes = units[i]:getHexes()
			for j = 1, uHexes:size() do
				hexes:insert(uHexes:at(j))
			end
		end
		return hexes
	end

	local destHex = mainDest.hex
	local targetIndex = 0

	while targetIndex < chainLength do
		local unit = battle:getUnitByPos(destHex, true)
		if not unit then break end

		local wouldResist = mechanics:wouldResist(unit)

		if alwaysHitFirst and targetIndex == 0 then
			table.insert(effectTarget, { unit = unit, hex = unit:getPosition() })
		end

		if wouldResist and targetIndex == 0 then
			if not alwaysHitFirst then
				table.insert(effectTarget, { unit = unit, hex = unit:getPosition() })
			end
			break
		elseif isEligible(unit) and not wouldResist then
			if not (alwaysHitFirst and targetIndex == 0) then
				table.insert(effectTarget, { unit = unit, hex = unit:getPosition() })
			end
		elseif isEligible(unit) and wouldResist then
			-- unit skipped, no resistance animation (H3 logic); don't advance targetIndex
			processedIds[unit] = true
			local hexes = buildPossibleHexes()
			if not hexes then break end
			destHex = destHex:getClosestTile(unit:unitSide(), hexes)
			goto continue
		else
			table.insert(effectTarget, {})
		end

		processedIds[unit] = true

		local hexes = buildPossibleHexes()
		if not hexes then break end
		destHex = destHex:getClosestTile(unit:unitSide(), hexes)

		targetIndex = targetIndex + 1
		::continue::
	end

	return effectTarget
end

return Script
